corehq/ex-submodules/toggle/models.py

```diff
@@ -25,8 +29,25 @@ class Toggle(Document):
     def save(self, **params):
         if ('_id' not in self._doc):
             self._doc['_id'] = generate_toggle_id(self.slug)
+        if ('_rev' not in self._doc):
+            self._rev = self._doc['_rev'] = uuid.uuid4().hex
+            rev_not_provided = True
+        else:
+            rev_not_provided = False
         self.last_modified = datetime.utcnow()
         super(Toggle, self).save(**params)
+        obj, created = SqlToggle.objects.get_or_create(
+            id=self._doc['_id'], defaults={'document': self, 'rev': self._rev}
+        )
```

Comment:
I think this can be done more efficiently (one query instead of (at least) two) with upsert
```
INSERT ...
ON CONFLICT (id) DO UPDATE ... WHERE rev = %s
RETURNING *
```
then looking at the number of rows returned to figure out if there was a resource conflict

Reply:
I thought this was looking to see if the `rev` was the same or not. Agree that doing it in one query would be great. I think the SQL way to do this is with a row lock using `select_for_update`

Reply:
upsert with a `WHERE clause` should handle the situation properly without locking

Reply:
The upsert I proposed will not affect any rows if the row was modified. In case it was not clear, the upsert also obviates the need for `obj.save()` below.

Reply:
Initially I wasn't very into the idea of putting custom SQL here, but I think that this is a fairly small function, and hopefully not one that would need to change much. https://github.com/dimagi/commcare-hq/pull/22690/commits/422d2c8732ea90a028eb4023fd67895dc40b28a2

I think it will make future model migrations cleaner as well.